### PR TITLE
Add PRODML write support

### PR DESCRIPTION
**Pull request by d-chambers**
## Summary

- add standalone PRODML 2.1-compatible HDF5 writing for one raw time-by-distance Patch
- derive acquisition, facility, units, sampling, and optional measurement metadata from Patch attrs and coordinates
- validate representability, preserve dimension order and dtype, round timestamps safely to microseconds, and replace existing targets only after validation
- clarify that full PRODML 2.3 EPC/XML packaging is outside this writer's scope

## Review

Claude ultrareview checked the branch against `dev`. Its one verified finding—invalid optional-measure units escaping the graceful omission path—was fixed with regression coverage.

## Validation

- `pytest -q tests/test_io/test_prodml` — 42 passed
- `pytest -q tests/test_io/test_io_core.py tests/test_io/test_common_io.py` — 1006 passed, 149 skipped
- `pre-commit run --all-files` — passed
- all 23 applicable official PRODML 2.3 DAS HDF5 examples still detect as PRODML 2.1

## Summary by CodeRabbit

* **New Features**
  * Added ability to write a single raw time-by-distance patch as a standalone PRODML 2.1-compatible HDF5 file via `dc.write`, with version-specific behavior and forwarded options support.
  * Added strict validation for patch shape, coordinates sampling/alignment, supported dtypes, units, timestamps, and required metadata (including UUIDs).
  * Implemented HDF5 string encoding utility for writing fixed-length UTF-8 strings.
* **Bug Fixes**
  * Improved PRODML metadata normalization/handling (including “UNKNOWN” units/type cleanup and safer version/fingerprint behavior).
* **Documentation**
  * Clarified current PRODML support level (write-only 2.1; no full 2.3 EPC/XML packaging).
* **Tests**
  * Expanded standalone PRODML write/read, precision rounding warnings, replacement semantics, and extensive negative/validation coverage.

## Changelog

- added: PRODML 2.1 write support for one raw time-by-distance patch as a standalone HDF5 file. Full PRODML 2.3 conformance requires EPC/XML packaging, which DASCore does not write.

### CODE REVIEW

**Comment by d-chambers** [comment]
@codex review

File: dascore/io/prodml/utils.py [hunk 1]
```diff
@@ -216,12 +223,244 @@ def _get_data_unit_and_type(node):
...
+    data_units = get_quantity_str(attrs.data_units) or "UNKNOWN"
+    string_values = (facility, service_company, data_units)
+    if not all(isinstance(x, str) and 0 < len(x) <= 64 for x in string_values):
+        msg = "ProdML facility, service company, and data unit must be String64 values."
+        raise PatchError(msg)
+
+    start_time, end_time = _iso_time(times[0]), _iso_time(times[-1])
+    output_rate = 1_000_000.0 / time_step_us
+    pulse_rate = _get_measure(attrs, "pulse_rate", "pulse_rate_units", "Hz")
+    pulse_rate = pulse_rate or (output_rate, "Hz")
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/io/prodml/utils.py]: Addressed in 29c17e0e. `PulseRate` now uses the optional-measure path and is omitted unless valid `pulse_rate` metadata and units are present; `OutputDataRate` remains derived only from the time cadence. Tests cover both omission and an explicit 50 Hz pulse rate.

File: dascore/io/prodml/utils.py [hunk 2]
```diff
@@ -216,12 +223,244 @@ def _get_data_unit_and_type(node):
...
+    try:
+        acquisition_id = str(UUID(acquisition_id))
+    except (ValueError, TypeError, AttributeError) as exc:
+        raise PatchError("ProdML acquisition_id must be a valid UUID.") from exc
+
+    facility = attrs.get("facility_id") or attrs.station or "UNKNOWN"
+    service_company = attrs.get("service_company_name") or "UNKNOWN"
+    data_units = get_quantity_str(attrs.data_units) or "UNKNOWN"
+    string_values = (facility, service_company, data_units)
+    if not all(isinstance(x, str) and 0 < len(x) <= 64 for x in string_values):
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/io/prodml/utils.py]: Addressed in 29c17e0e. String64 validation now checks `len(value.encode("utf-8"))` for each named HDF field and reports the offending field. The regression covers both 65 ASCII bytes and 33 two-byte characters.

File: dascore/io/prodml/utils.py [hunk 3]
```diff
@@ -216,12 +223,244 @@ def _get_data_unit_and_type(node):
...
+    times, time_step_us = _round_times_to_microseconds(patch.get_coord("time"))
+    distance_step, distance_units, start_locus = _get_distance_info(
+        patch.get_coord("distance")
+    )
+    attrs = patch.attrs
+    acquisition_id = str(attrs.acquisition_id or uuid4())
+    try:
+        acquisition_id = str(UUID(acquisition_id))
+    except (ValueError, TypeError, AttributeError) as exc:
+        raise PatchError("ProdML acquisition_id must be a valid UUID.") from exc
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/io/prodml/utils.py]: Addressed in 29c17e0e. The writer now preserves any non-empty `PatchAttrs.acquisition_id` verbatim and calls `uuid4()` only when the attribute is empty. `test_acquisition_id_is_preserved` covers `run-001`, while the missing-metadata test verifies that an absent ID receives a generated UUID.

File: dascore/io/prodml/utils.py [hunk 4]
```diff
@@ -39,6 +43,7 @@
     "GaugeLengthUnit": "gauge_length_units",
     "GaugeLengthUnits": "gauge_length_units",
     "GaugeLength.uom": "gauge_length_units",
+    "AcquisitionId": "acquisition_id",
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/io/prodml/utils.py:47]: Addressed in 29c17e0e. The root attribute map now restores `FacilityId` to `facility_id` and `ServiceCompanyName` to `service_company_name`; normalization collapses the writer singleton `FacilityId` array to a string while preserving multiple values as a tuple. Round-trip and station-fallback regression tests cover both paths.

File: dascore/io/prodml/utils.py [hunk 5]
```diff
@@ -216,12 +234,254 @@ def _get_data_unit_and_type(node):
...
         out["data_type"] = clean if clean in VALID_DATA_TYPES else ""
     return out
 
 
+# --- Writing ProdML files.
+
+
+def _fixed_string(value):
+    """Return a fixed-length HDF5 byte string."""
+    return np.bytes_(str(value).encode())
```
Comment (d-chambers) [inline dascore/io/prodml/utils.py]: This should go in the hdf5 utils

File: dascore/io/prodml/utils.py [hunk 6]
```diff
@@ -216,12 +234,254 @@ def _get_data_unit_and_type(node):
...
+
+
+def _fixed_strings(values):
+    """Return an array of fixed-length HDF5 byte strings."""
+    encoded = [str(value).encode() for value in values]
+    return np.asarray(encoded, dtype=f"S{max(map(len, encoded))}")
```
Comment (d-chambers) [inline dascore/io/prodml/utils.py]: Same for this. Probably better to combine to one function that optionally takes a string or sequence of strings. Use utils.iterate

File: dascore/io/prodml/utils.py [hunk 7]
```diff
@@ -216,12 +234,254 @@ def _get_data_unit_and_type(node):
...
+        msg = "ProdML writing requires a nonempty numeric distance coordinate."
+        raise PatchError(msg)
+    values = values.astype(np.float64)
+    if not np.all(np.isfinite(values)):
+        raise PatchError("ProdML distance coordinates must be finite.")
+    diffs = np.diff(values)
+    step = coord.step
+    if step is None:
+        step = diffs[0] if len(diffs) else np.nan
+    step = float(step)
```
Comment (d-chambers) [inline dascore/io/prodml/utils.py]: Just use the coord machinery here; this is probably redundant as coords already carry a step. You can also use require_evenly_sampled in get_coord to make this entire thing redundant.

File: dascore/io/prodml/utils.py [hunk 8]
```diff
@@ -216,12 +234,254 @@ def _get_data_unit_and_type(node):
...
+        msg = "ProdML distance coordinates must be increasing and evenly sampled."
+        raise PatchError(msg)
+    units = coord.unit_str
+    if not units:
+        raise PatchError("ProdML distance coordinates require length units.")
+    try:
+        convert_units(1.0, to_units="m", from_units=units)
+    except UnitError as exc:
+        msg = "ProdML distance coordinate units must represent length."
+        raise PatchError(msg) from exc
```
Comment (d-chambers) [inline dascore/io/prodml/utils.py]: Just assume in m if not provided. Again, use coordinate methods for converting units rather than redoing it here.

File: dascore/io/prodml/utils.py [hunk 9]
```diff
@@ -216,12 +234,254 @@ def _get_data_unit_and_type(node):
...
+    ratio = float(values[0] / step)
+    start_locus = int(np.round(ratio))
+    integer_tolerance = max(1e-9, abs(float(np.spacing(ratio))) * 4)
+    if not np.isclose(ratio, start_locus, rtol=0, atol=integer_tolerance):
+        msg = "ProdML requires distance_start / distance_step to be an integer."
```
Comment (d-chambers) [inline dascore/io/prodml/utils.py]: Is this right? Cant it be non-integer m? Please double check and simplify this.

File: dascore/io/prodml/utils.py [hunk 10]
```diff
@@ -216,12 +234,254 @@ def _get_data_unit_and_type(node):
...
+        raise PatchError(msg)
+    return step, units, start_locus
+
+
+def _get_measure(attrs, name, units_name, target_units):
```
Comment (d-chambers) [inline dascore/io/prodml/utils.py:320]: This smells of a general utility. Consider adding it to utils somewhere or double check that there isnt something like this already.

File: dascore/io/prodml/utils.py [hunk 11]
```diff
@@ -216,12 +234,254 @@ def _get_data_unit_and_type(node):
...
+        value = float(value)
+        if not units or not np.isfinite(value) or value <= 0:
+            return None
+        convert_units(value, to_units=target_units, from_units=units)
+    except (TypeError, ValueError, UnitError):
+        return None
+    return value, units
+
+
+def _iso_time(microseconds):
```
Comment (d-chambers) [inline dascore/io/prodml/utils.py]: This isnt needed. Cant we just use str(to_datetime64) then truncate the str?

File: dascore/io/prodml/utils.py [hunk 12]
```diff
@@ -216,12 +234,254 @@ def _get_data_unit_and_type(node):
...
+    value = np.datetime64(int(microseconds), "us")
+    return _fixed_string(f"{np.datetime_as_string(value, unit='us')}+00:00")
+
+
+def _prepare_prodml_write(spool):
+    """Validate input and prepare all values needed to write ProdML."""
+    patch = _get_single_patch(spool)
+    if patch.data.ndim != 2 or set(patch.dims) != {"time", "distance"}:
+        msg = "ProdML writing requires a 2D Patch with time and distance dimensions."
+        raise PatchError(msg)
```
Comment (d-chambers) [inline dascore/io/prodml/utils.py]: wrap these checks all up in _get_single_patch so it does all patch validation then either raises or returns the valid patch.

File: dascore/io/prodml/utils.py [hunk 13]
```diff
@@ -216,12 +234,254 @@ def _get_data_unit_and_type(node):
...
+    acquisition_id = str(attrs.acquisition_id or uuid4())
+
+    facility = attrs.get("facility_id") or attrs.station or "UNKNOWN"
+    service_company = attrs.get("service_company_name") or "UNKNOWN"
+    data_units = get_quantity_str(attrs.data_units) or "UNKNOWN"
+    string_values = {
+        "FacilityId": facility,
+        "ServiceCompanyName": service_company,
+        "RawDataUnit": data_units,
+    }
```
Comment (d-chambers) [inline dascore/io/prodml/utils.py]: This is ugly, maybe we can wrap it up into a pydantic model? EG subclass the current prodml patch attrs but add the extras/validation logic there.

File: dascore/io/prodml/utils.py [hunk 14]
```diff
@@ -216,12 +234,254 @@ def _get_data_unit_and_type(node):
...
+    return value, units
+
+
+def _iso_time(microseconds):
+    """Return an ISO-8601 UTC timestamp with microsecond precision."""
+    value = np.datetime64(int(microseconds), "us")
+    return _fixed_string(f"{np.datetime_as_string(value, unit='us')}+00:00")
+
+
+def _prepare_prodml_write(spool):
```
Comment (d-chambers) [inline dascore/io/prodml/utils.py]: This function is way too long. Please break into several sub functions (can be closures).

Comment (d-chambers) [reply dascore/io/prodml/utils.py]: You can also make a small class for this if it is better.

File: tests/test_io/test_prodml/test_prodml_write.py [hunk 15]
```diff
@@ -0,0 +1,461 @@
...
+        facility_id="facility-a",
+        service_company_name="service-a",
+        data_type="strain_rate",
+        data_units="m/s",
+    )
+
+
+def _decode(value):
+    """Decode one HDF5 string attribute."""
+    return str(unbyte(value))
```
Comment (d-chambers) [inline tests/test_io/test_prodml/test_prodml_write.py]: just inline this

File: tests/test_io/test_prodml/test_prodml_write.py [hunk 16]
```diff
@@ -0,0 +1,461 @@
...
+            assert set(file) == {"Acquisition"}
+            assert "sentinel" not in file.attrs
+            assert set(file["Acquisition"]) == {"Raw[0]"}
+
+
+class TestProdMLWriteRoundTrip:
+    """Representable Patch content should survive a write/read cycle."""
+
+    @pytest.mark.parametrize("dims", (("distance", "time"), ("time", "distance")))
+    def test_round_trip(self, dims, prodml_patch, tmp_path):
```
Comment (d-chambers) [inline tests/test_io/test_prodml/test_prodml_write.py:212]: Just compare the patches directly here; the Patch.__eq__ will handle the drudgery. (you can pop out some attrs if you need to before compare)

**Review by d-chambers — COMMENTED** [review]
Much of this code is overly verbose and duplicates existing functionality already part of dascore's objects or utils. Please address specific findings listed here and look for was to reduce line count. Ensure each helper function is more than a single line and actually needed.

File: dascore/io/prodml/utils.py (hunk 5 above)
Comment (d-chambers) [reply dascore/io/prodml/utils.py]: Addressed in a2ab81eb. The PRODML-local scalar encoder was removed and replaced by `encode_h5_strings` in `dascore.utils.hdf5`, so the HDF5 representation logic now lives with the shared backend utilities.

File: dascore/io/prodml/utils.py (hunk 6 above)
Comment (d-chambers) [reply dascore/io/prodml/utils.py]: Addressed in a2ab81eb. `encode_h5_strings` accepts either a string or a sequence and uses `iterate`; it returns an HDF5 scalar for a scalar input and a fixed-width ndarray for sequences, preserving the existing `FacilityId` and `Dimensions` shapes.

File: dascore/io/prodml/utils.py (hunk 7 above)
Comment (d-chambers) [reply dascore/io/prodml/utils.py]: Addressed in a2ab81eb. `_get_single_patch` now obtains both coordinates with `require_sorted=True` and `require_evenly_sampled=True`, eliminating the manual distance-difference/tolerance machinery. An explicit ascending check is retained because DASCore intentionally treats reverse-sorted coordinates as sorted; the regression suite now covers and rejects reversed distance.

File: dascore/io/prodml/utils.py (hunk 8 above)
Comment (d-chambers) [reply dascore/io/prodml/utils.py]: Addressed in a2ab81eb. Distance normalization now delegates to `coord.convert_units("m")`: it assigns meters when units are absent, converts compatible length units, and raises `UnitError` for incompatible units. The writer emits meters consistently, with a unitless-distance regression test.

File: dascore/io/prodml/utils.py (hunk 9 above)
Comment (d-chambers) [reply dascore/io/prodml/utils.py]: Addressed in a2ab81eb after checking the schema. `StartLocusIndex` is an integer locus index, not an integer-meter restriction. The simplified check only requires the physical start to align with the sampling grid; fractional coordinates such as start 0.5 m / step 0.5 m now have explicit successful round-trip coverage, while an unrepresentable start 0.5 m / step 2 m remains rejected.

File: dascore/io/prodml/utils.py (hunk 10 above)
Comment (d-chambers) [reply dascore/io/prodml/utils.py:320]: Checked in a2ab81eb. There is no existing DASCore utility combining a value/unit attribute pair with PRODML's format-specific 'omit when invalid' policy. `_get_measure` is substantive and reused for pulse rate, pulse width, and gauge length, so keeping it local avoids adding a nominally general API with only one consumer.

File: dascore/io/prodml/utils.py (hunk 11 above)
Comment (d-chambers) [reply dascore/io/prodml/utils.py]: Addressed in a2ab81eb. `_iso_time` was removed and the two timestamps are formatted inline from `datetime64[us]`. I did not route through `dc.to_datetime64`, because that normalizes through nanoseconds and would reintroduce overflow for the existing year-1600/year-2500 cases; `str(datetime64[us])` already provides exact microsecond precision.

File: dascore/io/prodml/utils.py (hunk 12 above)
Comment (d-chambers) [reply dascore/io/prodml/utils.py]: Addressed in a2ab81eb. `_get_single_patch` now owns cardinality, dimensions, nonempty supported dtype, even-sampling, and ascending-coordinate validation before returning the valid Patch. Representation-specific microsecond rounding and locus-grid conversion remain in their transform helpers so values are computed only once.

File: dascore/io/prodml/utils.py (hunk 13 above)
Comment (d-chambers) [reply dascore/io/prodml/utils.py]: Addressed in a2ab81eb with the smaller alternative. A Pydantic model would require aliases, byte-length validators, quantity handling, and exception translation for representation-only fields. `_get_write_metadata` now encapsulates the defaults and String64 validation directly, while the attribute builders consume its normalized values.

File: dascore/io/prodml/utils.py (hunk 14 above)
Comment (d-chambers) [reply dascore/io/prodml/utils.py]: Addressed in a2ab81eb. The former ~100-line preparation function is now a short coordinator, with substantive helpers for metadata normalization, Acquisition attributes, and raw/data/time array attributes. Together with coordinate and HDF5 utility reuse, the production implementation is five lines smaller while covering more edge cases.

File: tests/test_io/test_prodml/test_prodml_write.py (hunk 15 above)
Comment (d-chambers) [reply tests/test_io/test_prodml/test_prodml_write.py]: Addressed in a2ab81eb. The one-line `_decode` test helper was removed and its uses now call the existing `unbyte` utility directly.

File: tests/test_io/test_prodml/test_prodml_write.py (hunk 16 above)
Comment (d-chambers) [reply tests/test_io/test_prodml/test_prodml_write.py:212]: Addressed in a2ab81eb. The round-trip test now compares the patches directly after normalizing the two intentionally unstored core attrs (`tag` and `station`). It retains focused assertions for data dtype, coordinate units, and facility/service metadata because those are not covered by `Patch.__eq__`.
